src/Workspaces/Core/Portable/SymbolKey/SymbolKey.cs

```diff
@@ -75,114 +74,140 @@ namespace Microsoft.CodeAnalysis
-        {
-            return new SymbolKey(ToString(symbol, cancellationToken));
+            var result = reader.ReadSymbolKey();
+            Debug.Assert(reader.Position == symbolKey.Length);
+            return result;
         }
 
-        public static string ToString(ISymbol symbol, CancellationToken cancellationToken = default)
+        internal static string CreateString(ISymbol symbol, CancellationToken cancellationToken = default)
         {
-            using (var writer = SymbolKeyWriter.GetWriter(cancellationToken))
-            {
-                writer.WriteFirstSymbolKey(symbol);
-                return writer.CreateKey();
-            }
+            using var writer = SymbolKeyWriter.GetWriter(cancellationToken);
+            writer.WriteFormatVersion();
+            writer.WriteSymbolKey(symbol);
+            return writer.CreateKey();
         }
 
+        /// <summary>
+        /// Tries to resolve this <see cref="SymbolKey"/> in the given 
+        /// <paramref name="compilation"/> to a matching symbol.  <paramref name="resolveLocations"/>
+        /// should only be given <see langword="true"/> if the symbol was produced from a compilation
+        /// that has the exact same source as the compilation we're resolving against.  Otherwise
+        /// the locations resolved may not actually be correct in the final compilation.
+        /// </summary>
         public SymbolKeyResolution Resolve(
-            Compilation compilation,
-            bool ignoreAssemblyKey = false, bool resolveLocations = false,
-            CancellationToken cancellationToken = default)
+            Compilation compilation, bool ignoreAssemblyKey = false, bool resolveLocations = false, CancellationToken cancellationToken = default)
         {
-            return Resolve(
-                _symbolKeyData, compilation,
-                ignoreAssemblyKey, resolveLocations,
-                cancellationToken);
+            return ResolveString(_symbolKeyData, compilation, ignoreAssemblyKey, resolveLocations, cancellationToken);
         }
 
+        /// <summary>
+        /// Returns this <see cref="SymbolKey"/> encoded as a string.  This can be persisted
+        /// and used later with <see cref="SymbolKey(string)"/> to then try to resolve back
+        /// to the corresponding <see cref="ISymbol"/> in a future <see cref="Compilation"/>.
+        /// 
+        /// This string form is not guaranteed to be reusable across all future versions of 
+        /// Roslyn.  As such it should only be used for caching data, with the knowledge that
+        /// the data may need to be recomputed if the cached data can no longer be used.
+        /// </summary>
         public override string ToString()
             => _symbolKeyData;
 
-        private static IEnumerable<ISymbol> GetAllSymbols(SymbolKeyResolution info)
+        private static SymbolKeyResolution CreateSymbolInfo<TSymbol>(PooledArrayBuilder<TSymbol> symbols)
```

Comment:
Shouldn't this be called `CreateSymbolKeyResolution`?

Reply:
will do :)

Reply:
done.